File: app/src/main/java/com/akari/ppx/xp/hook/code/purity/CommentHook.java
package com.akari.ppx.xp.hook.code.purity;

import com.akari.ppx.common.utils.Utils;
import com.akari.ppx.common.utils.XSP;
import com.akari.ppx.xp.hook.code.SuperbHook;

import java.util.ArrayList;
import java.util.regex.Pattern;

import de.robv.android.xposed.XC_MethodHook;

import static com.akari.ppx.common.constant.Prefs.REMOVE_COMMENT;
import static com.akari.ppx.common.constant.Prefs.REMOVE_COMMENT_KEYWORDS;
import static com.akari.ppx.common.constant.Prefs.REMOVE_COMMENT_USERS;
import static de.robv.android.xposed.XposedHelpers.callMethod;
import static de.robv.android.xposed.XposedHelpers.getObjectField;

public class CommentHook extends SuperbHook {
	@Override
	protected void onHook(ClassLoader cl) {
		final ArrayList<String> keywords = new ArrayList<>(), users = new ArrayList<>();
		Utils.parseTextPlusAdd(keywords, XSP.gets(REMOVE_COMMENT_KEYWORDS));
		Utils.parseTextPlusAdd(users, XSP.gets(REMOVE_COMMENT_USERS));
		hookMethod(REMOVE_COMMENT, "com.sup.android.mi.feed.repo.response.a", "a", ArrayList.class, new XC_MethodHook() {
			@Override
			protected void beforeHookedMethod(MethodHookParam param) {
				ArrayList comments = (ArrayList) param.args[0];
				for (int i = 0; i < comments.size(); i++) {
					Object comment = null;
					String text, name;
					try {
						comment = callMethod(comments.get(i), "getReply");
						//Object reply2reply = getObjectField(comment, "reply");
					} catch (Error ignored) {
						comment = callMethod(comments.get(i), "getComment");
					} finally {
						name = (String) getObjectField(getObjectField(comment, "userInfo"), "name");
						text = (String) getObjectField(comment, "text");
					}
					for (String keyword : keywords) {
						if (Pattern.matches(keyword, text)) {
							comments.remove(i);
							break;
						}
					}
					for (String user : users) {
						if (Pattern.matches(user, name)) {
							comments.remove(i);
							break;
						}
					}
				}
			}
		});

	}
}

File: app/src/main/java/com/akari/ppx/xp/hook/purity/CommentHook.kt
@file:Suppress("unused")

package com.akari.ppx.xp.hook.purity

import com.akari.ppx.data.XPrefs
import com.akari.ppx.utils.*
import com.akari.ppx.xp.Init.commentResponseClass
import com.akari.ppx.xp.hook.SwitchHook
import java.util.regex.Pattern

class CommentHook : SwitchHook("remove_comments") {
    override fun onHook() {
        val (keywords, users) = listOf<String>(
            XPrefs("remove_comments_keywords"),
            XPrefs("remove_comments_users")
        ).map { it.splitByOr() }
        commentResponseClass!!.hookBeforeMethod("a", ArrayList::class.java) { param ->
            runCatching {
                val comments = param.args[0] as ArrayList<*>? ?: return@hookBeforeMethod
                comments.indices.reversed().forEach { i ->
                    run e@{
                        val comment = runCatching {
                            comments[i].callMethod("getReply")
                        }.getOrElse { comments[i].callMethod("getComment") }!!

                        fun checkPattern(list: ArrayList<String>, input: String?): Boolean {
                            input?.let {
                                list.forEach { s ->
                                    s.checkIf({ Pattern.matches(this, it) }) {
                                        comments.removeAt(i)
                                        return true
                                    }
                                }
                            }
                            return false
                        }
                        checkPattern(keywords, comment.getObjectFieldAs("text")).check(true) {
                            return@e
                        }
                        checkPattern(
                            users,
                            comment.getObjectField("userInfo")?.getObjectFieldAs("name")!!
                        ).check(true) {
                            return@e
                        }
                    }
                }
            }
        }
    }
}